cdmtaskservice/s3/client.py

```diff
@@ -0,0 +1,207 @@
+
+Note the client is *not threadsafe* as the underlying aiobotocore session is not threadsafe.
+"""
+
+from aiobotocore.session import get_session
+from botocore.config import Config
+from botocore.exceptions import EndpointConnectionError, ClientError
+from botocore.parsers import ResponseParserError
+import logging
+from typing import Any, Self
+
+
+class S3ObjectMeta:
+    """ Metadata about an object in S3 storage. """
+    
+    __slots__ = ["path", "e_tag", "part_size", "size"]
+    
+    def __init__(self, path: str, e_tag: str, size: int, part_size: int):
+        """
+        Create the object meta. This is not expected to be instantiated manually.
+        
+        path - the path of the object in s3 storage, including the bucket.
+        e-tag - the object e-tag.
+        size - the total size of the object
+        part_size - the part size used in a multipart upload, if relevant.
+        """
+        # input checking seems a little pointless. Could check that the path looks like a path
+        # and the e-tag looks like an e-tag, but since this module should be creating this
+        # class why bother. This is why access modifiers for methods are good
+        self.path = path
+        self.e_tag = e_tag
+        self.size = size
+        self.part_size = part_size
+        # could check that the part size makes sense given the e-tag... meh
+    
+    @property
+    def has_parts(self) -> bool:
+        """ Returns true if the object was uploaded as multipart. """
+        return len(self.e_tag) > 32
+
+    @property
+    def num_parts(self) -> int:
+        """ 
+        Returns the number of parts used in a multipart upload or 1 if the upload was not
+        multipart.
+        """
+        if len(self.e_tag) > 32:
+            return int(self.e_tag.split("-")[1])
+        return 1
+
+
+class S3Client:
+    """
+    The S3 client.
+    
+    Note the client is *not threadsafe* as the underlying aiobotocore session is not threadsafe.
+    """
+    
+    @classmethod
+    async def create(
+        cls,
+        endpoint_url: str,
+        access_key: str,
+        secret_key: str,
+        config: dict[str, Any] = None,
+        skip_connection_check: bool = False
+    ) -> Self:
+        """
+        Create the client.
+        
+        endpoint_url - the URL of the s3 endpoint.
+        access_key - the s3 access key.
+        secret_key - the s3 secret key.
+        config - Any client configuration options provided as a dictionary. See
+            https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html
+        skip_connetion_check - don't try to list_buckets while creating the client to check
+            the host and credentials are correct.
+        """
+        s3c = S3Client(endpoint_url, access_key, secret_key, config)
+        if not skip_connection_check:
+            async def list_buckets(client):
+                return await client.list_buckets()
```

Comment:
this will unlikely to raise permission issues right? It will just return an list in the event of no permission?

Reply:
No, it'll fail - see the tests

Reply:
The point of this is just to check the creds are valid at all. Since the service can theoretically read from any bucket, checking bucket perms doesn't make sense

Reply:
👍